### PR TITLE
[Models][GDN] Prevent D2H sync in `ChunkGatedDeltaRule`

### PR DESCRIPTION
**Pull request by lgeiger**
## Purpose

`ChunkGatedDeltaRule` currently causes 3 CUDA stream syncs:
1. `prepare_chunk_offsets` calls `.new_tensor([0])` causing H2D sync
        -> Fixed by using `.new_zeros(1)`
3. `prepare_chunk_indices` calls `.to(cu_seqlens)` causing H2D sync
        -> Fixed by using `.to(cu_seqlens, non_blocking=True)`
5. `prepare_chunk_indices` calls `triton.cdiv(prepare_lens(cu_seqlens), chunk_size).tolist()` causing D2H sync
        -> This is a less clean fix, passing through a `cu_seqlens_cpu` CPU tensor which removes the need for the D2H sync when calling `.tolist()`. The `.tolist()` call could be avoided by using `repeat_interleave` but it would still cause a D2H sync which wouldn't help much.

Screenshot 2026-03-16 at 18 11 30

## Accuracy Test

```
vllm serve Qwen/Qwen3.5-0.8B --limit-mm-per-prompt.video 0 --gpu-memory-utilization 0.95

python -m lm_eval --model local-completions   --model_args model=Qwen/Qwen3.5-0.8B,base_url=http://127.0.0.1:8000/v1/completions   --tasks gsm8k
```

**Before:**

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.3283|±  |0.0129|
|     |       |strict-match    |     5|exact_match|↑  |0.3245|±  |0.0129|
```

**After:**

```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.3389|±  | 0.013|
|     |       |strict-match    |     5|exact_match|↑  |0.3343|±  | 0.013|
```

## Performance Test

I verified in the torch profile that the syncs are gone and ran a benchmark showing 5% improved throughput and 13% lower TTFT.

```
vllm serve Qwen/Qwen3.5-2B --limit-mm-per-prompt.video 0

vllm bench serve --backend openai-chat --model Qwen/Qwen3.5-2B --endpoint /v1/chat/completions --dataset-name sharegpt --dataset-path ./ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts 1000
```

**main:**

```
============ Serving Benchmark Result ============
Successful requests:                     1000
Failed requests:                         0
Benchmark duration (s):                  18.79
Total input tokens:                      224742
Total generated tokens:                  201129
Request throughput (req/s):              53.21
Output token throughput (tok/s):         10702.52
Peak output token throughput (tok/s):    18491.00
Peak concurrent requests:                1000.00
Total token throughput (tok/s):          22661.53
---------------Time to First Token----------------
Mean TTFT (ms):                          3118.81
Median TTFT (ms):                        2957.29
P99 TTFT (ms):                           6047.54
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          75.65
Median TPOT (ms):                        46.48
P99 TPOT (ms):                           203.91
---------------Inter-token Latency----------------
Mean ITL (ms):                           35.70
Median ITL (ms):                         25.98
P99 ITL (ms):                            210.30
==================================================
```

**This PR:**

```
============ Serving Benchmark Result ============
Successful requests:                     1000
Failed requests:                         0
Benchmark duration (s):                  17.94
Total input tokens:                      224742
Total generated tokens:                  201572
Request throughput (req/s):              55.74
Output token throughput (tok/s):         11236.34
Peak output token throughput (tok/s):    18656.00
Peak concurrent requests:                1000.00
Total token throughput (tok/s):          23764.25
---------------Time to First Token----------------
Mean TTFT (ms):                          2705.37
Median TTFT (ms):                        2433.22
P99 TTFT (ms):                           5497.10
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          75.73
Median TPOT (ms):                        47.26
P99 TPOT (ms):                           199.94
---------------Inter-token Latency----------------
Mean ITL (ms):                           35.54
Median ITL (ms):                         25.90
P99 ITL (ms):                            207.59
==================================================
```

### CODE REVIEW

File: vllm/model_executor/layers/fla/ops/index.py [hunk 1]
```diff
@@ -15,27 +15,38 @@
...
-            for n in triton.cdiv(prepare_lens(cu_seqlens), chunk_size).tolist()
+            for n in triton.cdiv(
+                prepare_lens(cu_seqlens, cu_seqlens_cpu), chunk_size
+            ).tolist()
         ]
     )
-    return torch.stack([indices.eq(0).cumsum(0) - 1, indices], 1).to(cu_seqlens)
+    return torch.stack([indices.eq(0).cumsum(0) - 1, indices], 1).to(
+        cu_seqlens, non_blocking=True
+    )
```
[the remark this answers is not in the document]
Comment (lgeiger) [reply vllm/model_executor/layers/fla/ops/index.py:42]: This code would still cause an implicit sync

**Review by LucasWilkinson — COMMENTED** [review]
Overall looks good to me; would be good for @tdoublep to take a look; can you please provide E2E performance data?

**Review by ZJY0516 — CHANGES_REQUESTED** [review]
I think this involves too many changes for too little benefit.

File: vllm/model_executor/layers/fla/ops/index.py [hunk 2]
```diff
@@ -15,23 +15,35 @@
...
 
 
 @tensor_cache
 def prepare_chunk_offsets(cu_seqlens: torch.Tensor, chunk_size: int) -> torch.Tensor:
     return torch.cat(
-        [cu_seqlens.new_tensor([0]), triton.cdiv(prepare_lens(cu_seqlens), chunk_size)]
+        [cu_seqlens.new_zeros(1), triton.cdiv(prepare_lens(cu_seqlens), chunk_size)]
```
Comment (Jialin) [inline vllm/model_executor/layers/fla/ops/index.py:48]: minor: we could consider replace `torch.cat([zeros(1), X]).cumsum(-1)` with `F.pad(X.cumsum(-1), (1, 0))`

Comment (lgeiger) [reply vllm/model_executor/layers/fla/ops/index.py:48]: `.cat()` tends to have a bit less overhead compared to `.pad()` in simple cases like this. See #26486

But if you see a benefit in moving the cumsum before the padding, feel free to make a followup PR.

Comment (Jialin) [reply vllm/model_executor/layers/fla/ops/index.py:48]: Oh! Good to know!

Nah, I haven't compare perf among these 2 method, but originally though there might be less kernel calls involved.
